hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -234,27 +224,126 @@ private Pair<Integer, Boolean> submitPurgeKeysRequest(List<DeleteBlockGroupResul
         keyToUpdate.addAllKeyInfos(keyInfos);
         keysToUpdateList.add(keyToUpdate.build());
       }
+    }
 
-      if (!keysToUpdateList.isEmpty()) {
-        purgeKeysRequest.addAllKeysToUpdate(keysToUpdateList);
+    int purgeKeyIndex = 0, updateIndex = 0, renameIndex = 0;
+    int currSize = 0;
+    boolean batchCapacityReached;
+
+    while (purgeKeyIndex < purgeKeys.size() ||
+        updateIndex < keysToUpdateList.size() ||
+        (renameEntriesToBeDeleted != null && renameIndex < renameEntriesToBeDeleted.size())) {
+
+      int remainingRatisLimit = ratisLimit;
+      batchCapacityReached = false;
+      PurgeKeysRequest.Builder requestBuilder = PurgeKeysRequest.newBuilder();
+
+      if (snapTableKey != null) {
+        requestBuilder.setSnapshotTableKey(snapTableKey);
+      }
+      if (expectedPreviousSnapshotId != null) {
+        requestBuilder.setExpectedPreviousSnapshotID(
+            NullableUUID.newBuilder()
+                .setUuid(HddsUtils.toProtobuf(expectedPreviousSnapshotId))
+                .build());
+      }
+
+      currSize = requestBuilder.build().getSerializedSize();
+      remainingRatisLimit -= currSize;
+
+      // 3.1 Add purgeKeys
+      List<String> batchPurgeKeys = new ArrayList<>();
+      int estimatedPurgeKeysSize = 0;
+
+      while (purgeKeyIndex < purgeKeys.size()) {
+        String nextKey = purgeKeys.get(purgeKeyIndex);
+
+        List<String> temp = new ArrayList<>(batchPurgeKeys);
+        temp.add(nextKey);
+
+        OzoneManagerProtocolProtos.DeletedKeys deletedKeys = OzoneManagerProtocolProtos.DeletedKeys.newBuilder()
+            .setVolumeName("").setBucketName("").addAllKeys(temp).build();
+
+        PurgeKeysRequest.Builder tempBuilder = requestBuilder.clone();
+        tempBuilder.clearDeletedKeys().addDeletedKeys(deletedKeys);
```

Comment:
not an efficient code, to get batch, we serialize and check size, and create a batchList to operate. We can optimize this code avoiding another loop, and creating new map each time.